spec.html

```diff
@@ -112,51 +57,70 @@ <h1>Alarm Functions</h1>
   </emu-clause>
 
+  <emu-clause id="sec-addwaiter" aoid="AddWaiter">
+    <h1>AddWaiter ( _WL_, <del>_W_</del><ins>_waiterRecord_</ins> )</h1>
+    <p>The abstract operation AddWaiter takes two arguments, a WaiterList _WL_ and <del>an agent signifier _W_</del><ins>a Record _waiterRecord_</ins>. It performs the following steps:</p>
+    <emu-alg>
+      1. Assert: The calling agent is in the critical section for _WL_.
+      1. <del>Assert: _W_ is not on the list of waiters in any WaiterList.</del>
+      1. <ins>Let _inserted_ be *false*.</ins>
+      1. <ins>If _waiterRecord_.[[PromiseCapability]] is *null*, then</ins>
+        1. <ins>Assert: There is no record in _WL_ whose [[AgentSignifier]] field is _waiterRecord_.[[AgentSignifier]].</ins>
+        1. <ins>For each element _wr_ in _WL_, do</ins>
+          1. <ins>If _wr_.[[AgentSignifier]] is _waiterRecord_.[[AgentSignifier]], then</ins>
+            1. <ins>Insert _waiterRecord_ to immediately precede _wr_.</ins>
+            1. <ins>Set _inserted_ to *true*.</ins>
+      1. <ins>If _inserted_ is *false_, then</ins>
+        1. <ins>Append _waiterRecord_ as the last element of _WL_.</ins>
+    </emu-alg>
+  </emu-clause>
+
+  <emu-clause id="sec-removewaiter" aoid="RemoveWaiter">
+    <h1>RemoveWaiter ( _WL_, <del>_W_</del><ins>_waiterRecord_</ins> )</h1>
+    <p>The abstract operation RemoveWaiter takes two arguments, a WaiterList _WL_ and <del>an agent signifier _W_</del><ins>a Record _waiterRecord_</ins>. It performs the following steps:</p>
+    <emu-alg>
+      1. Assert: The calling agent is in the critical section for _WL_.
+      1. Assert: <del>_W_</del><ins>_waiterRecord</ins> is on the list of waiters in _WL_.
+      1. Remove <del>_W_</del><ins>_waiterRecord</ins> from the list of waiters in _WL_.
+    </emu-alg>
+  </emu-clause>
+
   <emu-clause id="sec-suspend" aoid="Suspend">
     <p>Change this function not to take a _timeout_ argument. Timeouts are now handled in the caller. (Not intended as a normative change.)</p>
     <h1>Suspend ( _WL_, _W_<del>, _timeout_</del> )</h1>
-    <p>The abstract operation Suspend takes <del>three</del><ins>two</ins> arguments, a <ins>Cluster</ins>WaiterList _WL_<del>,</del><ins>and</ins> an agent signifier _W_<del>, and a nonnegative, non-*NaN* Number _timeout_</del>. It performs the following steps:</p>
+    <p>The abstract operation Suspend takes <del>three</del><ins>two</ins> arguments, a WaiterList _WL_<del>,</del><ins>and</ins> an agent signifier _W_<del>, and a nonnegative, non-*NaN* Number _timeout_</del>. It performs the following steps:</p>
     <emu-alg>
       1. Assert: The calling agent is in the critical section for _WL_.
       1. Assert: _W_ is equal to AgentSignifier().
-      1. Assert: _W_ is on the list of waiters in _WL_.
+      1. Assert: <del>_W_ is on the list of waiters.</del><ins>There is a record</ins> in _WL_ <ins>whose [[AgentSignifier]] field is _W_ and whose [[PromiseCapability]] field is *null*.</ins>
```

Comment:
Nit: I think it'd be more idiomatic to use *undefined* rather than *null* to represent a missing promise capability (as well as a missing alarm).